### PR TITLE
Fix pending_calls metric to not lie

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/metrics/__init__.py [hunk 1]
```diff
@@ -158,18 +158,33 @@ def runUntilCurrentTimer(func):
...
+        num_pending = 0
+
+        # _newTimedCalls is one long list of *all* pending calls. Below loop
+        # is based off of impl of reactor.runUntilCurrent
+        for p in reactor._newTimedCalls:
+            if p.time > now:
+                break
+
+            if p.delayed_time > 0:
+                continue
```
Comment (NegativeMjark) [inline synapse/metrics/__init__.py]: It might be nice to have a more meaningful variable name than "p" here.

File: synapse/metrics/__init__.py [hunk 2]
```diff
@@ -158,18 +158,33 @@ def runUntilCurrentTimer(func):
 
     @functools.wraps(func)
     def f(*args, **kwargs):
-        pending_calls = len(reactor.getDelayedCalls())
+        now = reactor.seconds()
+        num_pending = 0
+
+        # _newTimedCalls is one long list of *all* pending calls. Below loop
+        # is based off of impl of reactor.runUntilCurrent
+        for p in reactor._newTimedCalls:
```
Comment (NegativeMjark) [inline synapse/metrics/__init__.py]: We might want to check that the _newTimedCalls attribute exists if we are rummaging around in twisted internals.

File: synapse/metrics/__init__.py [hunk 3]
```diff
@@ -158,18 +158,33 @@ def runUntilCurrentTimer(func):
...
+            if p.time > now:
+                break
+
+            if p.delayed_time > 0:
+                continue
+
+            num_pending += 1
+
+        num_pending += len(reactor.threadCallQueue)
```
Comment (NegativeMjark) [inline synapse/metrics/__init__.py]: Do we need to check if the "threadCallQueue" attribute exists?

Comment (erikjohnston) [inline synapse/metrics/__init__.py]: Done

**Comment by NegativeMjark** [comment]
LGTM

**Pull request merged**